building_sim_plugins/building_gazebo_plugins/include/crowd_simulator.hpp

```diff
@@ -0,0 +1,142 @@
+#ifndef BUILDING_SIM_COMMON__CROWD_SIMULATOR_GAZEBO_HPP
+#define BUILDING_SIM_COMMON__CROWD_SIMULATOR_GAZEBO_HPP
+
+#include <memory>
+#include <unordered_map>
+
+#include <ignition/math/Pose3.hh>
+
+#include <gazebo/common/Plugin.hh>
+#include <gazebo/common/UpdateInfo.hh>
+#include <gazebo/common/Events.hh>
+#include <gazebo/common/Mesh.hh>
+#include <gazebo/common/Skeleton.hh>
+#include <gazebo/common/SkeletonAnimation.hh>
+#include <gazebo/physics/Actor.hh>
+#include <gazebo/physics/physics.hh>
+
+#include <building_sim_common/crowd_simulator_common.hpp>
+
+
+namespace crowd_simulation_gazebo {
+
+using ObjectPtr = crowd_simulator::CrowdSimInterface::ObjectPtr;
+
+//================================================================
+
+template<typename... Args>
+using Task = std::function<bool(Args... args)>;
+
+template<typename TaskType>
+class TaskManager
+{
+public:
+  using TaskPtr = std::shared_ptr<TaskType>;
+
+  void AddTask(const TaskType& task);
+
+  template<typename... TaskArgs>
+  void RunAllTasks(TaskArgs... args);
+
+  size_t GetTasksCount();
+
+private:
+  std::list<TaskPtr> _tasks;
+};
+
+template<typename TaskType>
+void TaskManager<TaskType>::AddTask(const TaskType& task)
+{
+  this->_tasks.emplace_back(std::make_shared<TaskType>(task));
+}
+
+template<typename TaskType>
+template<typename... TaskArgs>
+void TaskManager<TaskType>::RunAllTasks(TaskArgs... args)
+{
+  std::list<TaskPtr> done;
+  for (const auto& task : this->_tasks)
+  {
+    if ((*task)(args...))
+    {
+      done.emplace_back(task);
+    }
+  }
+
+  for (const auto& task : done)
+  {
+    this->_tasks.remove(task);
+  }
+}
+
+template<typename TaskType>
+size_t TaskManager<TaskType>::GetTasksCount()
+{
+  return this->_tasks.size();
+}
```

Comment:
Is this TaskManager logic necessary? The ignition plugin seems to do fine without it?

Reply:
Not quite necessary, I can change it using traverse all the actors